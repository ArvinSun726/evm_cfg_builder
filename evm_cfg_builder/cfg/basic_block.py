class BasicBlock(object):

    def __init__(self):
        self.instructions = []
        # sons and fathers are dict
        # The key is the function hash
        # It allows to compute the VSA only
        # On a specific function, to separate
        # the merging
        self.sons = {}
        self.fathers = {}

        # List of function keys that reaches the BB
        self.reacheable = []

    def add_instruction(self, instruction):
        self.instructions += [instruction]

    def __repr__(self):
        return '<cfg BasicBlock@{:x}-{:x}>'.format(self.start.pc, self.end.pc)

    @property
    def start(self):
        return self.instructions[0]

    @property
    def end(self):
        return self.instructions[-1]

    def add_son(self, son, key):
        if not key in self.sons:
            self.sons[key] = []
        if son not in self.sons[key]:
            self.sons[key].append(son)

    def add_father(self, father, key):
        if not key in self.fathers:
            self.fathers[key] = []
        if father not in self.fathers[key]:
            self.fathers[key].append(father)

    def ends_with_jumpi(self):
        return self.end.name == 'JUMPI'

    def ends_with_jump_or_jumpi(self):
        return self.end.name in ['JUMP', 'JUMPI']

    def true_branch(self, key):
        assert(self.ends_with_jumpi())

        sons = [bb for bb in self.sons[key] if bb.start.pc != (self.end.pc+1)]
        assert(len(sons[key]) == 1)
        return sons[key][1]

    def false_branch(self, key):
        assert(self.ends_with_jumpi())

        sons = [bb for bb in self.sons[key] if bb.start.pc == (self.end.pc+1)]
        assert(len(sons) == 1)
        return sons[key][0]
