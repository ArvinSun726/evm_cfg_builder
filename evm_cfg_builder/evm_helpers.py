BASIC_BLOCK_END = ['STOP',
                   'SELFDESTRUCT',
                   'RETURN',
                   'REVERT',
                   'INVALID',
                   'SUICIDE',
                   'JUMP',
                   'JUMPI']

