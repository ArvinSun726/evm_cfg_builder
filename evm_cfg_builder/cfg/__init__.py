from .basic_block import BasicBlock
from .function import Function

__all__ = ["CFG", "BasicBlock", "Function"]

from ..known_hashes import known_hashes
from ..value_set_analysis import StackValueAnalysis

import re
from pyevmasm import disassemble_all, Instruction

BASIC_BLOCK_END = [
    'STOP',
    'SELFDESTRUCT',
    'RETURN',
    'REVERT',
    'INVALID',
    'SUICIDE',
    'JUMP',
    'JUMPI'
]

class CFG(object):
    """Implements the control flow graph (CFG) of an EVM bytecode.
    """

    def __init__(self, bytecode: bytes=None):
        """Initialize an EVM CFG.

        :param bytecode: The EVM bytecode
        :type bytecode: None, bytes
        """
        self._functions = dict()
        # __basic_blocks is a dict that matches
        # an address to the basic block
        # The address can be the first or the last
        # instructions 
        self._basic_blocks = dict()
        self._instructions = dict()

        assert(isinstance(bytecode, (type(None), bytes)))

        self._bytecode = bytecode
        self._analysis_complete = False

    @classmethod
    def from_byte_str(cls, bytecode: str) -> cls:
        return cls(bytes(bytecode.encode('charmap')))

    @classmethod
    def from_hex_str(cls, bytecode: str) -> cls:
        bytecode = bytes.fromhex(bytecode[2:])
        return cls(bytecode)

    @classmethod
    def from_bytes(cls, bytecode: bytes) -> cls:
        return cls(bytecode)

    @classmethod
    def from_hex_bytes(cls, bytecode: bytes) -> cls:
        return cls(bytes.fromhex(bytecode[2:].decode()))

    @property
    def analysis_complete(self) -> bool:
        return self._analysis_complete

    def __repr__(self) -> str:
        return "<CFG: {} Functions, {} Basic Blocks>".format(
            len(self.functions),
            len(self.basic_blocks)
        )

    @property
    def bytecode(self) -> bytes:
        return self._bytecode

    @property
    def basic_blocks(self) -> list:
        '''
        Return the list of basic_block
        '''
        bbs = self._basic_blocks.values()
        return list(set(bbs))

    @property
    def entry_point(self) -> BasicBlock:
        '''
        Return the entry point of the cfg (the basic block at 0x0)
        '''
        return self._basic_blocks[0]

    @property
    def functions(self) -> list:
        '''
        Return the list of functions
        '''
        return list(self._functions.values())

    @property
    def instructions(self) -> list:
        '''
        Return the list of instructions
        '''
        return list(self._instructions.values())

    def get_instruction_at(self, addr: int) -> Instruction:
        '''Return the instruction at the provided address.

        :param addr: Address of instruction
        :type addr: int
        '''
        return self._instructions.get(addr)

    def get_basic_block_at(self, addr: int) -> BasicBlock:
        '''Return the basic block at the provided address.

        The address is either the starting or ending instruction of the
        basic block.

        :param addr: Address of basic block start or end
        :type addr: int
        :return: BasicBlock, None -- the requested basic block
        '''
        return self._basic_blocks.get(addr)

    def get_function_at(self, addr: int) -> Function:
        '''Return the function at the provided address.

        :param addr: Address of the function
        :type addr: int
        :return: Function, None -- the requested function
        '''
        return self._functions.get(addr)

    def analyze(self):
        self.compute_basic_blocks()

        if self._basic_blocks:
            self.compute_functions(self._basic_blocks[0], True)
            self.add_function(Function(Function.DISPATCHER_ID, 0, self._basic_blocks[0], self))

        for function in self.functions:
            if function.hash_id in known_hashes:
                function.name = known_hashes[function.hash_id]

            vsa = StackValueAnalysis(
                self,
                function.entry,
                function.hash_id
            )
            bbs = vsa.analyze()

            function.basic_blocks = [self._basic_blocks[bb] for bb in bbs]

            if function.hash_id != Function.DISPATCHER_ID:
                function.check_payable()
                function.check_view()
                function.check_pure()

        self._analysis_complete = True

    def clear(self):
        self._functions = dict()
        self._basic_blocks = dict()
        self._instructions = dict()
        self._bytecode = bytes()
        self._analysis_complete = False

    def remove_metadata(self):
        '''
            Init bytecode contains metadata that needs to be removed
            see http://solidity.readthedocs.io/en/v0.4.24/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
        '''
        if self._bytecode is None:
            return

        self.bytecode = re.sub(
            bytes(r'\xa1\x65\x62\x7a\x7a\x72\x30\x58\x20[\x00-\xff]{32}\x00\x29'.encode('charmap')),
            b'',
            self.bytecode
        )

    def compute_basic_blocks(self):
        '''
            Split instructions into BasicBlock
        Args:
            self: CFG
        Returns:
            None
        '''
        # Do nothing if bytecode is None
        if self._bytecode is None:
            return

        # Do nothing if basic_blocks already exists
        if self._basic_blocks:
            return

        bb = BasicBlock()

        for instruction in disassemble_all(self.bytecode):
            self._instructions[instruction.pc] = instruction

            if instruction.name == 'JUMPDEST':
                # JUMPDEST indicates a new BasicBlock. Set the end pc
                # of the current block, and switch to a new one.
                if bb._instructions:
                    self._basic_blocks[bb.end.pc] = bb

                bb = BasicBlock()

                self._basic_blocks[instruction.pc] = bb

            bb.add_instruction(instruction)

            if bb.start.pc == instruction.pc:
                self._basic_blocks[instruction.pc] = bb

            if bb.end.name in BASIC_BLOCK_END:
                self._basic_blocks[bb.end.pc] = bb
                bb = BasicBlock()

    def compute_functions(self, block: BasicBlock, is_entry_block:bool=False):
        function_start, function_hash = is_jump_to_function(block)

        if(function_start):
            new_function = Function(
                function_hash,
                function_start,
                self._basic_blocks[function_start],
                self
            )

            self._functions[function_start] = new_function

            if block.ends_with_jumpi():
                false_branch = self._basic_blocks[block.end.pc + 1]
                self.compute_functions(false_branch)

            return

        elif is_entry_block:
            if block.ends_with_jumpi():
                false_branch = self._basic_blocks[block.end.pc + 1]
                self.compute_functions(false_branch)

    def add_function(self, func: Function):
        assert isinstance(func, Function)
        self._functions[func._start_addr] = func

    def compute_simple_edges(self, key):
        for bb in self._basic_blocks.values():

            if bb.end.name == 'JUMPI':
                dst = self._basic_blocks[bb.end.pc + 1]
                bb.add_outgoing_basic_block(dst, key)
                dst.add_incoming_basic_block(bb, key)

            # A bb can be split in the middle if it has a JUMPDEST
            # Because another edge can target the JUMPDEST
            if bb.end.name not in BASIC_BLOCK_END:
                try:
                    dst = self._basic_blocks[bb.end.pc + 1 + bb.end.operand_size]
                except KeyError:
                    continue
                assert dst.start.name == 'JUMPDEST'
                bb.add_outgoing_basic_block(dst, key)
                dst.add_incoming_basic_block(bb, key)

    def compute_reachability(self, entry_point: int, key):
        bbs_saw = [entry_point]

        bbs_to_explore = [entry_point]
        while bbs_to_explore:
            bb = bbs_to_explore.pop()
            for son in bb.outgoing_basic_blocks(key):
                if not son in bbs_saw:
                    bbs_saw.append(son)
                    bbs_to_explore.append(son)

        for bb in bbs_saw:
            bb.reacheable.append(key)

        # clean son/fathers that are created by compute_simple_edges
        # but are not reacheable
        for bb in self._basic_blocks.values():
            if not bb in bbs_saw:
                if key in bb._incoming_basic_blocks:
                    bb._incoming_basic_blocks.pop(key)
                if key in bb._outgoing_basic_blocks:
                    bb._outgoing_basic_blocks.pop(key)

    def output_to_dot(self, base_filename: str):

        with open('{}{}.dot'.format(base_filename, 'FULL_GRAPH'), 'w') as f:
            f.write('digraph{\n')
            for basic_block in self.basic_blocks:
                instructions = ['{}:{}'.format(hex(ins.pc),
                                               str(ins)) for ins in basic_block.instructions]
                instructions = '\n'.join(instructions)

                f.write('{}[label="{}"]\n'.format(basic_block.start.pc, instructions))

                for son in basic_block.all_incoming_basic_blocks:
                    f.write('{} -> {}\n'.format(basic_block.start.pc, son.start.pc))

            f.write('\n}')

def is_jump_to_function(block: BasicBlock) -> bool:
    '''
        Heuristic:
        Recent solc version add a first check if calldatasize <4 and jump in fallback
    Args:
        block (BasicBlock)
    Returns:
        (int): function hash, or None
    '''

    has_calldata_size = False
    last_pushed_value = None
    previous_last_pushed_value = None
    for i in block.instructions:
        if i.name == 'CALLDATASIZE':
            has_calldata_size = True

        if i.name.startswith('PUSH'):
            previous_last_pushed_value = last_pushed_value
            last_pushed_value = i.operand

    if i.name == 'JUMPI' and has_calldata_size:
        return last_pushed_value, -1

    if i.name == 'JUMPI' and previous_last_pushed_value:
        return last_pushed_value, previous_last_pushed_value

    return None, None
