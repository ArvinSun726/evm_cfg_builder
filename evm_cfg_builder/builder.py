import sys
import re
import binascii

from pyevmasm import disassemble_all

from value_set_analysis import StackValueAnalysis
from cfg import split_instructions, find_functions

def remove_metadata(bytecode):
    '''
        Init bytecode contains metadata that needs to be removed
        see http://solidity.readthedocs.io/en/v0.4.24/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
    '''
    if len(bytecode) < 86:
        return bytecode

    metadata = bytecode[-86::]
    if re.match('a165627a7a72305820[0-9a-z]{64}0029', metadata):
        return bytecode[:-86]
    return bytecode


def get_info(bytecode):

    bytecode = remove_metadata(bytecode)
    instructions = list(disassemble_all(binascii.unhexlify(bytecode))) # todo use generator
    basic_blocks = split_instructions(instructions)

    vsa = StackValueAnalysis(basic_blocks)
    vsa.analyze()

    find_functions(basic_blocks[0], True)

    output_to_dot(basic_blocks)

def output_to_dot(basic_blocks):
    with open('test.dot', 'w') as f:
        f.write('digraph{\n')

        for basic_block in basic_blocks:
            instructions = ['{}:{}'.format(hex(ins.pc), str(ins)) for ins in basic_block.instructions]
            instructions = '\n'.join(instructions)
            f.write('{}[label="{}"]\n'.format(basic_block.start.pc, instructions))
            for son in basic_block.sons:
                f.write('{} -> {}\n'.format(basic_block.start.pc, son.start.pc))
        f.write('\n}')



if __name__ == '__main__':

    filename = sys.argv[1]

    with open(filename) as f:
        get_info(f.read().replace('\n',''))
