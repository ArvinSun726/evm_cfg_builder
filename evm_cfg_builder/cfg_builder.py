import binascii
import re
import sys

from pyevmasm import disassemble_all

from .cfg import compute_instructions, find_functions, Function
from .evm_helpers import create_dicts_from_basic_blocks
from .known_hashes import known_hashes
from .value_set_analysis import StackValueAnalysis

DISPATCHER_ID = -2
FALLBACK_ID = -1


def remove_metadata(bytecode):
    '''
        Init bytecode contains metadata that needs to be removed
        see http://solidity.readthedocs.io/en/v0.4.24/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
    '''
    return re.sub(
        r'\xa1\x65\x62\x7a\x7a\x72\x30\x58\x20[\x00-\xff]{32}\x00\x29',
        '',
        bytecode
    )


def get_info(bytecode):
    instructions = disassemble_all(bytecode)
    basic_blocks = compute_instructions(instructions)
    basic_blocks_as_dict, nodes_as_dict = create_dicts_from_basic_blocks(
        basic_blocks
    )

    functions = find_functions(basic_blocks[0], basic_blocks_as_dict, True)
    dispatcher = Function(-2, 0, basic_blocks_as_dict[0])
    functions = [dispatcher] + functions

    for function in functions:
        if function.hash_id == FALLBACK_ID:
            function.name = '_fallback'
        elif function.hash_id == DISPATCHER_ID:
            function.name = '_dispatcher'
        else:
            if function.hash_id in known_hashes:
                function.name = known_hashes[function.hash_id]

    for function in functions:
        vsa = StackValueAnalysis(
            function.entry,
            basic_blocks_as_dict,
            nodes_as_dict,
            function.hash_id
        )
        bbs = vsa.analyze()

        function.basic_blocks = [basic_blocks_as_dict[bb] for bb in bbs]

        function.check_payable()
        function.check_view()
        function.check_pure()

    return functions

def output_to_dot(functions):
    for function in functions:
        function.output_to_dot('test_')

def main():
    filename = sys.argv[1]

    with open(filename) as f:
        bytecode = f.read().replace('\n','')
        bytecode = binascii.unhexlify(bytecode)
        bytecode = remove_metadata(bytecode)
        functions = get_info(bytecode)
        print('End of analysis')
        for function in functions:
            print(function)
        output_to_dot(functions)


if __name__ == '__main__':
    main()
