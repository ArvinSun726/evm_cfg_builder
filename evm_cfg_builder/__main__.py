from .cfg_builder import main

main()