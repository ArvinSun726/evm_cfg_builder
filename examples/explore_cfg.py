import pprint
from evm_cfg_builder.cfg import CFG

with open('token-runtime0.5.evm') as f:
    runtime_bytecode = f.read()

cfg = CFG(runtime_bytecode)

for basic_block in cfg.basic_blocks:
    print('{} -> {}'.format(basic_block, basic_block.all_outgoing_basic_blocks))
