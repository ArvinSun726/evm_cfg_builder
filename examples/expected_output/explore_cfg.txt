<cfg BasicBlock@0-c> -> [<cfg BasicBlock@d-40>, <cfg BasicBlock@af-b3>]
<cfg BasicBlock@109-11c> -> [<cfg BasicBlock@11d-135>, <cfg BasicBlock@136-143>]
<cfg BasicBlock@11d-135> -> [<cfg BasicBlock@136-143>]
<cfg BasicBlock@136-143> -> []
<cfg BasicBlock@144-14b> -> [<cfg BasicBlock@14c-14f>, <cfg BasicBlock@150-18e>]
<cfg BasicBlock@14c-14f> -> []
<cfg BasicBlock@150-18e> -> [<cfg BasicBlock@5b9-6aa>]
<cfg BasicBlock@18f-1a8> -> []
<cfg BasicBlock@1a9-1b0> -> [<cfg BasicBlock@1b1-1b4>, <cfg BasicBlock@1b5-1bd>]
<cfg BasicBlock@1b1-1b4> -> []
<cfg BasicBlock@1b5-1bd> -> [<cfg BasicBlock@6ab-6b0>]
<cfg BasicBlock@1be-1d3> -> []
<cfg BasicBlock@1d4-1db> -> [<cfg BasicBlock@1dc-1df>, <cfg BasicBlock@1e0-23e>]
<cfg BasicBlock@1dc-1df> -> []
<cfg BasicBlock@1e0-23e> -> [<cfg BasicBlock@6b1-77c>]
<cfg BasicBlock@23f-258> -> []
<cfg BasicBlock@259-260> -> [<cfg BasicBlock@261-264>, <cfg BasicBlock@265-299>]
<cfg BasicBlock@261-264> -> []
<cfg BasicBlock@265-299> -> [<cfg BasicBlock@94b-962>]
<cfg BasicBlock@29a-2af> -> []
<cfg BasicBlock@2b0-2b7> -> [<cfg BasicBlock@2b8-2bb>, <cfg BasicBlock@2bc-2c4>]
<cfg BasicBlock@2b8-2bb> -> []
<cfg BasicBlock@2bc-2c4> -> [<cfg BasicBlock@963-975>]
<cfg BasicBlock@2c5-2e0> -> []
<cfg BasicBlock@2e1-2e8> -> [<cfg BasicBlock@2e9-2ec>, <cfg BasicBlock@2ed-341>]
<cfg BasicBlock@2e9-2ec> -> []
<cfg BasicBlock@2ed-341> -> [<cfg BasicBlock@976-99a>]
<cfg BasicBlock@342-357> -> []
<cfg BasicBlock@358-35f> -> [<cfg BasicBlock@360-363>, <cfg BasicBlock@364-398>]
<cfg BasicBlock@360-363> -> []
<cfg BasicBlock@364-398> -> [<cfg BasicBlock@99b-9e3>]
<cfg BasicBlock@399-3ae> -> []
<cfg BasicBlock@3af-3b6> -> [<cfg BasicBlock@3b7-3ba>, <cfg BasicBlock@3bb-3c3>]
<cfg BasicBlock@3b7-3ba> -> []
<cfg BasicBlock@3bb-3c3> -> [<cfg BasicBlock@9e4-a33>]
<cfg BasicBlock@3c4-3e7> -> [<cfg BasicBlock@3e9-3f1>]
<cfg BasicBlock@3e9-3f1> -> [<cfg BasicBlock@3f2-403>, <cfg BasicBlock@404-417>]
<cfg BasicBlock@3f2-403> -> [<cfg BasicBlock@3e9-3f1>]
<cfg BasicBlock@404-417> -> [<cfg BasicBlock@418-430>, <cfg BasicBlock@431-43e>]
<cfg BasicBlock@41-4b> -> [<cfg BasicBlock@4c-56>, <cfg BasicBlock@144-14b>]
<cfg BasicBlock@418-430> -> [<cfg BasicBlock@431-43e>]
<cfg BasicBlock@431-43e> -> []
<cfg BasicBlock@43f-446> -> [<cfg BasicBlock@447-44a>, <cfg BasicBlock@44b-489>]
<cfg BasicBlock@447-44a> -> []
<cfg BasicBlock@44b-489> -> [<cfg BasicBlock@a82-acd>]
<cfg BasicBlock@48a-4a3> -> []
<cfg BasicBlock@4a4-4ab> -> [<cfg BasicBlock@4ac-4af>, <cfg BasicBlock@4b0-504>]
<cfg BasicBlock@4ac-4af> -> []
<cfg BasicBlock@4b0-504> -> [<cfg BasicBlock@bdb-c61>]
<cfg BasicBlock@4c-56> -> [<cfg BasicBlock@57-61>, <cfg BasicBlock@1a9-1b0>]
<cfg BasicBlock@505-51a> -> []
<cfg BasicBlock@51b-56a> -> [<cfg BasicBlock@56b-572>, <cfg BasicBlock@5b1-5b8>]
<cfg BasicBlock@56b-572> -> [<cfg BasicBlock@573-585>, <cfg BasicBlock@586-593>]
<cfg BasicBlock@573-585> -> [<cfg BasicBlock@5b1-5b8>]
<cfg BasicBlock@57-61> -> [<cfg BasicBlock@62-6c>, <cfg BasicBlock@1d4-1db>]
<cfg BasicBlock@586-593> -> [<cfg BasicBlock@594-5a7>]
<cfg BasicBlock@594-5a7> -> [<cfg BasicBlock@594-5a7>, <cfg BasicBlock@5a8-5b0>]
<cfg BasicBlock@5a8-5b0> -> [<cfg BasicBlock@5b1-5b8>]
<cfg BasicBlock@5b1-5b8> -> [<cfg BasicBlock@c9-ec>]
<cfg BasicBlock@5b9-6aa> -> [<cfg BasicBlock@18f-1a8>]
<cfg BasicBlock@62-6c> -> [<cfg BasicBlock@6d-77>, <cfg BasicBlock@259-260>]
<cfg BasicBlock@6ab-6b0> -> [<cfg BasicBlock@1be-1d3>]
<cfg BasicBlock@6b1-77c> -> [<cfg BasicBlock@77d-781>, <cfg BasicBlock@782-788>]
<cfg BasicBlock@6d-77> -> [<cfg BasicBlock@78-82>, <cfg BasicBlock@2b0-2b7>]
<cfg BasicBlock@77d-781> -> [<cfg BasicBlock@782-788>]
<cfg BasicBlock@782-788> -> [<cfg BasicBlock@789-78c>, <cfg BasicBlock@78d-84f>]
<cfg BasicBlock@78-82> -> [<cfg BasicBlock@83-8d>, <cfg BasicBlock@2e1-2e8>]
<cfg BasicBlock@789-78c> -> []
<cfg BasicBlock@78d-84f> -> [<cfg BasicBlock@850-8d9>, <cfg BasicBlock@8da-94a>]
<cfg BasicBlock@83-8d> -> [<cfg BasicBlock@8e-98>, <cfg BasicBlock@358-35f>]
<cfg BasicBlock@850-8d9> -> [<cfg BasicBlock@8da-94a>]
<cfg BasicBlock@8da-94a> -> [<cfg BasicBlock@23f-258>]
<cfg BasicBlock@8e-98> -> [<cfg BasicBlock@99-a3>, <cfg BasicBlock@3af-3b6>]
<cfg BasicBlock@94b-962> -> [<cfg BasicBlock@29a-2af>]
<cfg BasicBlock@963-975> -> [<cfg BasicBlock@2c5-2e0>]
<cfg BasicBlock@976-99a> -> [<cfg BasicBlock@342-357>]
<cfg BasicBlock@99-a3> -> [<cfg BasicBlock@a4-ae>, <cfg BasicBlock@43f-446>]
<cfg BasicBlock@99b-9e3> -> [<cfg BasicBlock@399-3ae>]
<cfg BasicBlock@9e4-a33> -> [<cfg BasicBlock@a34-a3b>, <cfg BasicBlock@a7a-a81>]
<cfg BasicBlock@a34-a3b> -> [<cfg BasicBlock@a3c-a4e>, <cfg BasicBlock@a4f-a5c>]
<cfg BasicBlock@a3c-a4e> -> [<cfg BasicBlock@a7a-a81>]
<cfg BasicBlock@a4-ae> -> [<cfg BasicBlock@af-b3>, <cfg BasicBlock@4a4-4ab>]
<cfg BasicBlock@a4f-a5c> -> [<cfg BasicBlock@a5d-a70>]
<cfg BasicBlock@a5d-a70> -> [<cfg BasicBlock@a5d-a70>, <cfg BasicBlock@a71-a79>]
<cfg BasicBlock@a71-a79> -> [<cfg BasicBlock@a7a-a81>]
<cfg BasicBlock@a7a-a81> -> [<cfg BasicBlock@3c4-3e7>]
<cfg BasicBlock@a82-acd> -> [<cfg BasicBlock@ace-ad1>, <cfg BasicBlock@ad2-bda>]
<cfg BasicBlock@ace-ad1> -> []
<cfg BasicBlock@ad2-bda> -> [<cfg BasicBlock@48a-4a3>]
<cfg BasicBlock@af-b3> -> []
<cfg BasicBlock@b4-bb> -> [<cfg BasicBlock@bc-bf>, <cfg BasicBlock@c0-c8>]
<cfg BasicBlock@bc-bf> -> []
<cfg BasicBlock@bdb-c61> -> [<cfg BasicBlock@505-51a>]
<cfg BasicBlock@c0-c8> -> [<cfg BasicBlock@51b-56a>]
<cfg BasicBlock@c62-c62> -> []
<cfg BasicBlock@c9-ec> -> [<cfg BasicBlock@ee-f6>]
<cfg BasicBlock@d-40> -> [<cfg BasicBlock@41-4b>, <cfg BasicBlock@b4-bb>]
<cfg BasicBlock@ee-f6> -> [<cfg BasicBlock@f7-108>, <cfg BasicBlock@109-11c>]
<cfg BasicBlock@f7-108> -> [<cfg BasicBlock@ee-f6>]
